pkg/controllers/hpareplicassyncer/hpa_scale_ref_worker.go

```diff
@@ -0,0 +1,134 @@
+		return nil
+	}
+
+	switch event.kind {
+	case addLabelEvent:
+		err = r.addHPALabelToScaleRef(event.hpa)
+	case deleteLabelEvent:
+		err = r.deleteHPALabelFromScaleRef(event.hpa)
+	default:
+		klog.Errorf("Found invalid key when reconciling hpa scale ref: %+v", key)
+		return nil
+	}
+
+	if err != nil {
+		klog.Errorf("reconcile scale ref failed: %+v", err)
+	}
+	return err
+}
+
+func (r *HPAReplicasSyncer) addHPALabelToScaleRef(hpa *autoscalingv2.HorizontalPodAutoscaler) error {
+	scaleRef, mapping, err := r.getScaleRef(hpa)
+	if err != nil {
+		return err
+	}
+
+	newScaleRef := scaleRef.DeepCopy()
+	util.MergeLabel(newScaleRef, util.ResourceScaledByHPALabel, hpa.GetName())
+
+	// use patch is better than update, when modification occur after get, patch can still success while update can not
+	if err = r.patchScaleRef(mapping.Resource, scaleRef, newScaleRef); err != nil {
+		return err
+	}
+
+	klog.Infof("add hpa labels to %+v success", hpa.Spec.ScaleTargetRef)
+	return nil
+}
+
+func (r *HPAReplicasSyncer) deleteHPALabelFromScaleRef(hpa *autoscalingv2.HorizontalPodAutoscaler) error {
+	scaleRef, mapping, err := r.getScaleRef(hpa)
+	if err != nil {
+		return err
+	}
+
+	newScaleRef := scaleRef.DeepCopy()
+	util.RemoveLabels(newScaleRef, util.ResourceScaledByHPALabel)
+
+	// use patch is better than update, when modification occur after get, patch can still success while update can not
+	if err = r.patchScaleRef(mapping.Resource, scaleRef, newScaleRef); err != nil {
+		return err
+	}
+
+	klog.Infof("delete hpa labels from %+v success", hpa.Spec.ScaleTargetRef)
+	return nil
+}
+
+func (r *HPAReplicasSyncer) getScaleRef(hpa *autoscalingv2.HorizontalPodAutoscaler) (*unstructured.Unstructured, *meta.RESTMapping, error) {
+	targetGVK := schema.FromAPIVersionAndKind(hpa.Spec.ScaleTargetRef.APIVersion, hpa.Spec.ScaleTargetRef.Kind)
+	mapping, err := r.RESTMapper.RESTMapping(targetGVK.GroupKind(), targetGVK.Version)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to recognize scale ref resource: %+v", err)
+	}
+
+	scaleRef, err := r.DynamicClient.Resource(mapping.Resource).Namespace(hpa.Namespace).
+		Get(context.Background(), hpa.Spec.ScaleTargetRef.Name, metav1.GetOptions{})
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Infof("scale ref resource is not found (%+v), skip processing", hpa.Spec.ScaleTargetRef)
+			return nil, nil, nil
+		}
+		return nil, nil, fmt.Errorf("failed to find scale ref resource: %+v, err: %+v", scaleRef, err)
+	}
+
+	return scaleRef, mapping, nil
+}
+
+func (r *HPAReplicasSyncer) patchScaleRef(resource schema.GroupVersionResource, scaleRef, newScaleRef *unstructured.Unstructured) error {
+	patchBytes, err := helper.GenMergePatch(scaleRef, newScaleRef)
+	if err != nil {
+		return fmt.Errorf("failed to gen merge patch: %+v", err)
+	}
+	if len(patchBytes) == 0 {
+		klog.Infof("hpa labels not exist, skip deleting (%+v)", resource)
```

Comment:
I'm a little confused about this message. Both adding and deleting actions call this method, why empty patchBytes only matches deleting action?

Reply:
nice finding, at the begining there are two same code in each function, I refactored the code to just one function but mistakenly overlooked this log content

Reply:
my refactor didn't really reduce code lines, I plan to change it back